snow/validators/state.go

```diff
@@ -165,3 +175,74 @@ func (n *noValidators) GetCurrentValidatorSet(ctx context.Context, _ ids.ID) (ma
 	height, err := n.GetCurrentHeight(ctx)
 	return nil, height, err
 }
+
+type cachedState struct {
+	State
+
+	// Activate caching of individual validator sets after this time.
+	//
+	// This enables pre-Granite code to call GetWarpValidatorSet without
+	// generating the full validator set for each P-chain height.
+	//
+	// After Granite, it is expected that the P-chain height will be frozen for
+	// the duration of an epoch, so generating the full set is not expensive as
+	// long as it is cached.
+	activation time.Time
+
+	// Caches validators for all subnets at various heights.
+	// Key: height
+	// Value: mapping height -> subnet ID -> validator set
```

Comment:
```suggestion
	// Value: mapping subnet ID -> validator set
```

...or am I misunderstanding something?